### ISSUE TITLE
docker_service only finds docker-compose.yml on local runs (CentOS/RHEL)

### ISSUE THREAD

**Issue by andrefernandes** [ansible/ansible-modules-core#4085]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

Module: docker_service
##### ANSIBLE VERSION

Ansible 2.1 (installed with "pip install"). Results from "ansible --version":

```
ansible 2.1.0.0
  config file =
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

N/A
##### OS / ENVIRONMENT

Controller node: CentOS 7.2
Remote node: RHEL 7.1
##### SUMMARY

When using playbooks the module only works with "docker-compose.yml" files if ran locally. Remote runs only work with inline definitions.
##### STEPS TO REPRODUCE

Take this simple playbook below (file named "web-deploy.yml"):

```
- name: Nginx deployment
  hosts: '{{ target }}'
  tasks:
  - docker_image:
      name: nginx
  - docker_service:
      project_name: nginx_sample
      project_src: '.'
```
##### EXPECTED RESULTS

It works ok when ran locally:

```
ansible-playbook web-deploy.yml --extra-vars "target=local"

PLAY [Nginx deployment] ********************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [docker_image] ************************************************************
ok: [localhost]

TASK [docker_service] **********************************************************
ok: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0

```
##### ACTUAL RESULTS

Fails when 'target' is a remote host. Please notice that 'docker_image' task runs ok, so I did figure out all the remote python dependencies.

```
ansible-playbook web-deploy.yml --extra-vars "target=dados-t" -u goodsshuser

PLAY [Nginx deployment] ********************************************************

TASK [setup] *******************************************************************
ok: [myserver]

TASK [docker_image] ************************************************************
ok: [myserver]

TASK [docker_service] **********************************************************
fatal: [myserver]: FAILED! => {"changed": false, "failed": true, "msg": "Configuration error - \n        Can't find a suitable configuration file in this directory or any\n        parent. Are you in the right directory?\n\n        Supported filenames: docker-compose.yml, docker-compose.yaml\n        "}

NO MORE HOSTS LEFT *************************************************************
        to retry, use: --limit @web-deploy.retry

PLAY RECAP *********************************************************************
myserver  : ok=2    changed=0    unreachable=0    failed=1

```
##### EXTRA INFO

The remote RHEL node had its python dependencies installed with the following commands:

```
pip install PyYAML
pip install docker-py
pip install docker-compose
```

The remote RHEL node runs Docker's docker engine (1.11), **not** the one in RHEL distro, but this should not be an issue.

**Comment by zagyi** [comment]
I'have also hit this problem. My quick workaround is to upload the docker-compose.yml files as follows:

```
- set_fact: compose_dir=/var/tmp/ansible-docker-service

- file: 
    path: "{{ compose_dir }}"
    state: directory

- name: Upload docker-compose files
  synchronize: 
    src: "{{ item }}"
    dest: "{{ compose_dir }}"
  with_items:
    - gitlab
    - other_prjs
```

And then refer to the project so:

```
- name: Start GitLab
  docker_service:
    project_src: "{{ compose_dir }}/gitlab"
```

Not nice, but works for now.

**Comment by chouseknecht** [comment]
Not sure this is a bug report as much as a feature request. But, seems like a reasonable request. I'll see if we can perhaps use the same logic as the file module.

**Comment by kiview** [comment]
I use the same workaround as @zagyi. But it's interesting to hear, that this is indeed a missing feature and not some misconfiguration of the module.

I'd be happy very if this would be implemented. If this is a low priority feature for the core developers, I could try to help you out if you like?

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by chouseknecht** [comment]
@kiview @zagyi @andrefernandes 

Feature PR out there for review. See [PR 1695](https://github.com/ansible/ansible/pull/16954)

**Comment by kiview** [comment]
Great work @chouseknecht, this feature together with https://github.com/ansible/ansible-modules-core/pull/4307 makes Ansible the perfect docker provisioning tool, thanks a lot for your effort :)

**Comment by chouseknecht** [comment]
wontfix

This got shot down on review. It duplicates too much of the copy plugin code. It is convenient, but it creates a maintenance headache going forward. It will overtime duplicate problems we experience with the copy module today around permissions. Overtime we'll end up adding all the copy module parameters to the docker_service module. 

It was decided that the better solution is to simply call the copy module directly from the playbook. And of course, an Ansible role could be used to combine calls to the copy module and docker_service module.

**Comment by kiview** [comment]
Bummer, but I totally understand the reasoning behind the decision.

**Comment by sw360cab** [comment]
to summarize will we able to deal with the issue when PR 1695 may be merged?

**Comment by kiview** [comment]
It seems there won't be a merge of this concrete solution.

**Comment by sw360cab** [comment]
ok @kiview , sorry but I got lost. I just encountered the same issue using ansible (can't make images work on target host). 
So actually this feauture would not be supported by **docker-service**, or am I missing smt?

**Comment by kiview** [comment]
I think it won't be supported in a way, that the file will be copied to the external host. Maybe there could be a solution, which will read the docker-compose file locally and act similar as to how the module would behave if the module was configured directly inside the playbook/role.

**Comment by sw360cab** [comment]
Well, probably I need to read again the doc about **docker_service**, but in my experience using this plugin in a playbook to configure a role on a target host (aka running docker project on target host), was not possible (as the issue is stating). At this point this plugins looks useless in a real prod scenario. 
But I am sure I am missing smt.

**Comment by chouseknecht** [comment]
@sw360cab Like all other cloud modules, if the source files do not exist on the deployment node, they have to be created or copied there. This module works the same way that the s3 module works, for example.

**Comment by sw360cab** [comment]
Thank you @chouseknecht. As I said my concern were mainly because of my ignorance.
I'll give try to what you explained.

**Comment by sean-abbott** [comment]
I think the documentation should make this more explicit.  It should say "project_src must be on the inventory host" or something.

**Comment by torvitas** [comment]
I've configured the remote docker host to be accessible via https and then configured the docker_service tasks to talk to that docker host. (https://docs.ansible.com/ansible/docker_service_module.html#options)
I also had to set connection: local in the playbook.

This way ansible uses the local docker-compose.yml and executes on the remote docker host.

It were awesome if I could do the docker host configuration elsewhere and make ansible pick it up via the inventory configuration.

**Comment by mvitale1989** [comment]
Kind of counter intuitive, but i understand why `project_dir` is expected to be local to the inventory host. 

As a solution, as @chouseknecht suggested, I created the [docker-compose-target][1] role to automate copy&deploy of a local docker-compose project, which produces the behaviour expected by @andrefernandes in this issue.

[1]:https://github.com/mvitale1989/docker-compose-target
